core/network/src/service.rs

```diff
@@ -540,19 +558,17 @@ fn run_thread<B: BlockT + 'static>(
 						version <= protocol::CURRENT_VERSION as u8
 						&& version >= protocol::MIN_VERSION as u8
 					);
-					let _ = protocol_sender.send(FromNetworkMsg::PeerConnected(peer_id, debug_info));
-				}
-				Ok(Async::Ready(Some(NetworkServiceEvent::ClosedCustomProtocol { peer_id, debug_info, .. }))) => {
-					let _ = protocol_sender.send(FromNetworkMsg::PeerDisconnected(peer_id, debug_info));
-				}
-				Ok(Async::Ready(Some(NetworkServiceEvent::CustomMessage { peer_id, message, .. }))) => {
-					let _ = protocol_sender.send(FromNetworkMsg::CustomMessage(peer_id, message));
+					protocol.on_peer_connected(peer_id, debug_info);
```

Comment:
Have you considered keeping those `on_` methods private, and enqueuing messages instead? That way there is only one place where the `protocol` state can change, and it's after [this call to `poll`](https://github.com/paritytech/substrate/pull/2475/files#diff-9c1b34b73c6680d46fc07cae4591547fR364). 

With the current setup, where everything runs in one task, it doesn't make much difference I guess(and is sort of equivalent to the `select` over two separate channels), and if you wanted to make those tasks run independently I think you wouldn't be able to share the entire `protocol` here, and sharing a sender would be easier. 

Could be done as part of a follow-up in the light of the other comment...

Reply:
The purpose of this PR is to remove threads and tasks.